cdap-api/src/main/java/io/cdap/cdap/api/app/Application.java

```diff
@@ -32,4 +32,18 @@
    * @param context Used to access the environment, application configuration, and application (deployment) arguments
    */
   void configure(ApplicationConfigurer configurer, ApplicationContext<T> context);
+
+  /**
+   * Returns if application supports config upgrade or not.
+   */
+  boolean isUpgradeAllowed();
+
+  /**
+   * Upgrades given configuration.
+   *
+   * @param config Configuration, most likely older, to be upgraded.
+   * @param applicationUpgradeContext Used to access methods helpful for upgrading a config such as get latest plugin
+   * versions.
+   */
+  ConfigUpgradeResult<T> upgradeConfig(T config, ApplicationUpgradeContext applicationUpgradeContext) throws Exception;
```

Comment:
thinking about this a little more, this API means the config class before and after the upgrade has to be the same. This means the app can't change the class at all, which seems overly restrictive. I'm also not sure how the classloading will work, since the config passed in comes from the old artifact while the config returned comes from the new artifact. 

I wonder if it would be better to just pass in the config string?

Reply:
The signature better be:

```
ApplicationUpdateResult<T> update(ApplicationUpdateContext context);
```

Also, better to throw more specific exception instead of `Exception`.

Regarding to Albert question, The ApplicationUpdateContext should have at least two methods:

```
// Get the old config as an object. The platform would perform the json deserialization based on the provided type.
// This is for the case where the new app has the same / compatible / old config class.
<C extends Config> C getConfig(Type configType);

// Get the config as a json string.
String getConfigAsString();
```

Reply:
Good suggestion. Changed accordingly. Done.